Microsoft Azure transcription service improvements

wsmirnow opened the pull request:
The transcription service should write results into workspace. Minor improvements in logging and error handling are also done.

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

lkiesow commented on modules/transcription-service-microsoft-azure/src/main/java/org/opencastproject/transcription/microsoftazure/MicrosoftAzureTranscriptionService.java:
```diff
@@ -500,37 +486,37 @@ void recognizeContinuous(SpeechRecognizer speechRecognizer, String jobId, String
-          logger.error(String.format("Could not write to tmp file: %s", text));
+          writeTranscriptToFile(text, jobId);
+        } catch (NotFoundException ex) {
+          logger.debug("Speech recognized before session start. "
+              + "Transcription file not created before as it should be.");
```
I don't quite understand what you are trying to say here, and think the second sentence is broken?
It's also not matching the error from the next line.

wsmirnow replied:
MS Azure text to speech API is event driven. You push the audio stream and Azure will throw events at you. First event should be `sessionStarted`, where we will create an empty captions file. After that some of `recognized` events may appear. But if we got an `recognized` event before `sessionStarted`, we want to throw an error (and log it). May be 

> Transcription file not created before, as it should be.

sounds better?

wsmirnow replied:
```suggestion
          logger.debug("Speech recognized before session started. "
              + "Transcription file not created before, as it should be.");
```

lkiesow commented on docs/guides/admin/docs/modules/microsoftazuretranscripts.md:
````diff
@@ -251,8 +251,25 @@ to the mediapackages and republishes it. Copy it into a new file under
+sudo apt-get install libgstreamer1.0-0 gstreamer1.0-plugins-base gstreamer1.0-plugins-good gstreamer1.0-plugins-bad gstreamer1.0-plugins-ugly
+```
+
+For RedHat based systems run:
````
```suggestion
On Red Hat based systems, run:
```

lkiesow commented on docs/guides/admin/docs/modules/microsoftazuretranscripts.md:
```diff
@@ -251,8 +251,25 @@ to the mediapackages and republishes it. Copy it into a new file under
 
+### Step 5: Install gstreamer
+
+Microsoft Speech-to-text API uses gstreamer to transcode audio stream.
```
That's unfortunate. Is it a dependency of `com.microsoft.cognitiveservices.speech:client-sdk`?

lkiesow commented on docs/guides/admin/docs/modules/microsoftazuretranscripts.md:
```diff
@@ -251,8 +251,25 @@ to the mediapackages and republishes it. Copy it into a new file under
+
+Microsoft Speech-to-text API uses gstreamer to transcode audio stream.
+Therefor gstreamer must be installed on your Opencast admin node.
+On Debian based system you can do it by running:
```
```suggestion
The Microsoft speech-to-text client uses GStreamer to transcode audio stream.
Therefore, GStreamer must be installed on your Opencast admin node.
On Debian based systems, you can do that by running:
```

lkiesow commented on modules/transcription-service-microsoft-azure/src/main/java/org/opencastproject/transcription/microsoftazure/MicrosoftAzureTranscriptionService.java:
```diff
@@ -559,33 +545,39 @@ else if (CancellationReason.Error == e.getReason()) {
+    return workspace.putInCollection(TRANSCRIPT_COLLECTION, getTranscriptFileName(jobId),
+        new ByteArrayInputStream(transcript.getBytes(StandardCharsets.UTF_8)));
+  }
+  private void writeTranscriptToFile(String text, String jobId) throws NotFoundException, IOException {
```
```suggestion
  }

  private void writeTranscriptToFile(String text, String jobId) throws NotFoundException, IOException {
```

lkiesow reviewed (changes requested):
This is hard to test, but changes generally look good with a few minor change requests.

wsmirnow committed:
Apply suggestions from code review

Co-authored-by: Lars Kiesow <lkiesow@uos.de>

wsmirnow committed:
Microsoft Azure transcription service improvements

The transcription service should write results into workspace. Minor improvements in logging and error handling are also done.

wsmirnow commented:
I've solved the comments and rebased to current r/12.x.

lkiesow commented:
> That's unfortunate. Is it [GStreamer] a dependency of `com.microsoft.cognitiveservices.speech:client-sdk`?

Do you know the answer to that question?

gregorydlogan commented on modules/transcription-service-microsoft-azure/src/main/java/org/opencastproject/transcription/microsoftazure/MicrosoftAzureTranscriptionService.java:
```diff
@@ -499,37 +487,37 @@ void recognizeContinuous(SpeechRecognizer speechRecognizer, String jobId, String
-          writeToTmpFile(text, jobId);
-        } catch (IOException ioException) {
-          logger.error(String.format("Could not write to tmp file: %s", text));
+          writeTranscriptToFile(text, jobId);
```
Why are we creating the file, then recognizing, and only then writing to the file?  Why not just create the file when it's time to write?

gregorydlogan commented on modules/transcription-service-microsoft-azure/src/main/java/org/opencastproject/transcription/microsoftazure/MicrosoftAzureTranscriptionService.java:
```diff
@@ -538,52 +526,60 @@ else if (ResultReason.NoMatch == e.getResult().getReason()) {
     speechRecognizer.sessionStopped.addEventListener((s, e) -> {
       try {
-        logger.info("Session stopped");
+        logger.trace("Transcriptiong job {} session stopped", jobId);
```
Transcription~g~

gregorydlogan commented on docs/guides/admin/docs/modules/microsoftazuretranscripts.md:
````diff
@@ -251,8 +251,25 @@ to the mediapackages and republishes it. Copy it into a new file under
+On Debian based systems, you can do that by running:
+
+```shell
+sudo apt-get install libgstreamer1.0-0 gstreamer1.0-plugins-base gstreamer1.0-plugins-good gstreamer1.0-plugins-bad gstreamer1.0-plugins-ugly
````
This installs a whole bunch of things, some of which might be [problematic](https://gstreamer.freedesktop.org/modules/gst-plugins-ugly.html).  If these are hard requirements then we should be including a warning here with a link to the page above.  If they're *not* then I would suggest dropping `-ugly` at the very least.

wsmirnow replied:
Here is what the [Speech SDK documentation](https://learn.microsoft.com/en-us/azure/cognitive-services/speech-service/how-to-use-codec-compressed-audio-input-streams?tabs=linux%2Ccentos%2Cjava-android%2Cterminal&pivots=programming-language-csharp) say about:

> The Speech SDK can use [GStreamer](https://gstreamer.freedesktop.org/) to handle compressed audio

In theory I've testet this module w/o gstreamer installed and it failed. Then I've started to install gstreamer-plugins-base, this also failed. This game was go on until all documented plugins are installed. So yes, the speech SDK needs at least gstreamer1-plugins-ugly-free (RedHat) or gstreamer1.0-plugins-ugly (Debian)!

I'm not the author of the transcription-service opencast module nor Speech SDK. This are my findings with a clean opencast installation. In my opinion we should document the dependencies instead of let the user run into the issue.

wsmirnow replied:
> > That's unfortunate. Is it [GStreamer] a dependency of `com.microsoft.cognitiveservices.speech:client-sdk`?
> 
> Do you know the answer to that question?

Yes, it is. This is documented [here](https://learn.microsoft.com/en-us/azure/cognitive-services/speech-service/how-to-use-codec-compressed-audio-input-streams?tabs=linux%2Ccentos%2Cjava-android%2Cterminal&pivots=programming-language-csharp).

gregorydlogan replied:
Fair enough.  I don't have to like it for it to be a hard dependency :D

Let's add it to the dos and be done with it.

wsmirnow commented:
To solve the merge conflicts will take more time as rewrite. Please go ahead with review of #4441.
